packages/service-library/src/servicelib/background_task.py

```diff
@@ -56,21 +58,43 @@ async def start_periodic_task(
         )
 
 
+async def cancel_task(
+    task: asyncio.Task,
+    *,
+    timeout: float | None,
+    cancellation_attempts: int = _MAX_TASK_CANCELLATION_ATTEMPTS,
+) -> None:
+    """Reliable task cancellation. Some libraries will just hang without
+    cancelling the task. It is important to retry the operation to provide
+    a timeout in that situation to avoid forever pending tasks.
+
+    :param task: task to be canceled
+    :param timeout: total duration (in seconds) to wait before giving
+        up the cancellation. If None it waits forever.
+    :raises TryAgain: raised if cannot cancel the task.
+    """
+    async for attempt in AsyncRetrying(
+        stop=stop_after_attempt(cancellation_attempts), reraise=True
```

Comment:
no before_sleep_log?
also what will be reraised? the TryAgain exception? Maybe add a specific PeriodicTaskCancellationError or similar?

Reply:
No log there, it just creates clutter from my point of view. The current logging is sufficent, from my point of view.

Let's go for a custom exception.